pkg/fleet-manager/application_controller.go

```diff
@@ -0,0 +1,384 @@
+	app := &applicationapi.Application{}
+	if err := a.Get(ctx, req.NamespacedName, app); err != nil {
+		if apiserrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get application %s", req.NamespacedName)
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+	log = log.WithValues("application", klog.KObj(app))
+
+	patchHelper, err := patch.NewHelper(app, a.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to init patch helper for application %s", req.NamespacedName)
+	}
+
+	defer func() {
+		patchOpts := []patch.Option{}
+		if err := patchHelper.Patch(ctx, app, patchOpts...); err != nil {
+			reterr = utilerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch application %s", req.NamespacedName)})
+		}
+	}()
+
+	// Add finalizer if not exist to void the race condition.
+	if !controllerutil.ContainsFinalizer(app, ApplicationFinalizer) {
+		controllerutil.AddFinalizer(app, ApplicationFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	// Define object key for fleet based on the current policy's destination
+	fleetKey := client.ObjectKey{
+		Namespace: app.Namespace,
+		// there only one fleet
+		Name: app.Spec.SyncPolicy[0].Destination.Fleet,
+	}
+	fleet := &fleetapi.Fleet{}
+	// Retrieve fleet object based on the defined fleet key
+	if err := a.Client.Get(ctx, fleetKey, fleet); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("fleet does not exist", "fleet", fleetKey)
+			return ctrl.Result{RequeueAfter: RequeueAfter}, nil
+		}
+
+		// Log error and requeue request if error occurred during fleet retrieval
+		log.Error(err, "failed to find fleet", "fleet", fleetKey)
+		return ctrl.Result{}, err
+	}
+
+	// label the fleet
+	if fleet.GetLabels() == nil {
+		fleet.SetLabels(make(map[string]string))
+	}
+	// todo: there is a bug here if more than one application refer the same fleet: the new one will override the old
+	//  but it is needed to reconcile application when the fleet' cluster is all joined
+	if fleet.GetLabels()[ApplicationLabel] != app.Name {
+		labels := fleet.GetLabels()
+		labels[ApplicationLabel] = app.Name
+		fleet.SetLabels(labels)
+		err := a.Update(ctx, fleet)
+		if err != nil {
+			log.Error(err, "unable to label fleet", "fleet", fleet.Name)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Handle deletion reconciliation loop.
+	if app.DeletionTimestamp != nil {
+		return a.reconcileDelete(ctx, app, fleet)
+	}
+
+	// Handle normal loop.
+	return a.reconcile(ctx, app, fleet)
+}
+
+func (a *ApplicationManager) reconcile(ctx context.Context, app *applicationapi.Application, fleet *fleetapi.Fleet) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	if result, err := a.reconcileInitializeParameters(ctx, app); err != nil || result.RequeueAfter > 0 {
+		log.Error(err, "failed to reconcileInitializeParameters")
+		return result, err
+	}
+
+	if result, err := a.reconcileSyncResources(ctx, app, fleet); err != nil || result.RequeueAfter > 0 {
```

Comment:
reconcileSyncResources/syncResources

i think reconcile and sync have similar meaning

Reply:
reconcileApplicationResource